File: apps/api/internal/galleryactions/gallery_actions_test.go
package galleryactions

import (
	"encoding/json"
	"strings"
	"testing"
)

func valid() []Descriptor {
	return []Descriptor{{Version: 1, ID: "acme.remove-bg", Label: "Remove background", AcceptedMediaTypes: []string{"image/png"}, SchemaRevision: 1, Fields: []Field{{ID: "quality", Kind: "number", Label: "Quality", Min: ptr(1), Max: ptr(10)}, {ID: "format", Kind: "select", Label: "Format", Choices: []Choice{{ID: "png", Label: "PNG"}}}}}}
}
func ptr(v float64) *float64 { return &v }
func TestValidateDescriptors(t *testing.T) {
	for _, tc := range []struct {
		name   string
		mutate func([]Descriptor)
	}{
		{"valid", func([]Descriptor) {}}, {"unsupported version", func(d []Descriptor) { d[0].Version = 2 }}, {"unsupported field", func(d []Descriptor) { d[0].Fields[0].Kind = "html" }}, {"unbounded number", func(d []Descriptor) { d[0].Fields[0].Min = nil; d[0].Fields[0].Max = nil }}, {"callback cannot be an id", func(d []Descriptor) { d[0].ID = "https://evil.invalid" }},
	} {
		t.Run(tc.name, func(t *testing.T) {
			d := valid()
			tc.mutate(d)
			err := ValidateDescriptors(d)
			if tc.name == "valid" && err != nil {
				t.Fatal(err)
			}
			if tc.name != "valid" && err == nil {
				t.Fatal("accepted unsafe descriptor")
			}
		})
	}
}
func TestSubmissionDigestIsPayloadBoundAndStable(t *testing.T) {
	a := SubmissionDigest("s", "retry-1", "{\"x\":1}")
	if a != SubmissionDigest("s", "retry-1", "{\"x\":1}") {
		t.Fatal("not stable")
	}
	if a == SubmissionDigest("s", "retry-1", "{\"x\":2}") {
		t.Fatal("not payload bound")
	}
}

func TestStrictGalleryWireContract(t *testing.T) {
	good := `{"version":1,"id":"synthetic.adjust","label":"Adjust","accepted_media_types":["image/png"],"schema_revision":1,"fields":[{"id":"images","kind":"images","label":"Images","min":0,"max":2,"dynamic":true,"choices":[]},{"id":"amount","kind":"number","label":"Amount","min":1,"max":3,"step":0.5,"default":1.5}]}`
	var d Descriptor
	if err := Decode([]byte(good), &d); err != nil {
		t.Fatal(err)
	}
	if err := ValidateDescriptors([]Descriptor{d}); err != nil {
		t.Fatal(err)
	}
	encoded, err := json.Marshal(d)
	if err != nil {
		t.Fatal(err)
	}
	if err = Decode(encoded, &d); err != nil {
		t.Fatal(err)
	}
	if err = ValidateDescriptors([]Descriptor{d}); err != nil {
		t.Fatal("roundtrip", err)
	}
	for _, tc := range []struct{ name, from, to string }{
		{"unknown descriptor", `"version":1`, `"html":"x","version":1`},
		{"duplicate key", `"version":1`, `"version":1,"version":1`},
		{"null", `"default":1.5`, `"default":null`},
		{"unaligned default", `"default":1.5`, `"default":1.25`},
		{"zero step", `"step":0.5`, `"step":0`},
		{"unsafe step", `"step":0.5`, `"step":1e-20`},
		{"bad bounds", `"max":3`, `"max":0`},
		{"fractional images", `"min":0`, `"min":0.5`},
		{"unbounded images", `"max":2`, `"max":101`},
		{"unknown field", `"dynamic":true`, `"callback_url":"https://evil"`},
		{"missing fields", `,"fields":`, `,"wrong":`},
		{"missing version", `"version":1,`, ``},
	} {
		t.Run(tc.name, func(t *testing.T) {
			var bad Descriptor
			err := Decode([]byte(strings.Replace(good, tc.from, tc.to, 1)), &bad)
			if err == nil {
				err = ValidateDescriptors([]Descriptor{bad})
			}
			if err == nil {
				t.Fatal("accepted invalid wire contract")
			}
		})
	}
	for _, value := range []string{`{"images":[],"amount":1.5}`, `{"images":[],"amount":1.25}`, `{"images":[],"amount":1.5,"extra":true}`, `{"images":["foreign"],"amount":1.5}`} {
		var values map[string]any
		if err = Decode([]byte(value), &values); err != nil {
			t.Fatal(err)
		}
		err = ValidateValues(d.Fields, values)
		if (err == nil) != (value == `{"images":[],"amount":1.5}`) {
			t.Fatalf("unexpected validation %s: %v", value, err)
		}
	}
}

File: apps/web/src/lib/gallery-actions.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import test from "node:test";
import {
  defaultGalleryActionValues,
  eligibleGalleryActions,
  validateGalleryActionSchema,
  validateGalleryActionValues,
} from "./gallery-actions.ts";
const upload = { id: "upload-1", content_type: "image/png" } as any;
const fields = [
  {
    id: "images",
    kind: "images",
    label: "Images",
    choices: [{ id: "a", label: "A", upload_id: "upload-a" }],
    min: 0,
    max: 2,
    dynamic: true,
  },
  { id: "amount", kind: "number", label: "Amount", min: 1, max: 3, step: 0.5, default: 1.5 },
  { id: "confirm", kind: "boolean", label: "Confirm", default: true },
  { id: "format", kind: "select", label: "Format", choices: [{ id: "png", label: "PNG" }] },
];
const allowed = new Set(["upload-a", "preview"]);
const descriptor = {
  version: 1,
  id: "synthetic.adjust",
  label: "Adjust",
  accepted_media_types: ["image/png"],
  schema_revision: 1,
  fields,
};
test("wire descriptors match the generated host contract and preserve installation identity", () => {
  const action = { installation_id: "app-1", descriptor };
  assert.deepEqual(eligibleGalleryActions([action, action, null, []], upload), [action]);
  for (const change of [
    { version: 2 },
    { id: "photo:adjust" },
    { callback_url: "https://invalid" },
    { fields: null },
    { accepted_media_types: ["image/*"] },
    { accepted_media_types: ["image/png", "image/png"] },
  ])
    assert.deepEqual(
      eligibleGalleryActions([{ ...action, descriptor: { ...descriptor, ...change } }], upload),
      [],
    );
});
test("schema rejects null, unknown properties, unsupported controls, invalid defaults and bounds", () => {
  const schema = validateGalleryActionSchema({ revision: 1, fields, preview: "preview" }, allowed);
  assert.ok(schema);
  assert.equal(validateGalleryActionValues(schema, defaultGalleryActionValues(schema)), undefined);
  const bad = [
    null,
    [],
    { revision: 1, fields: null },
    { revision: 1, fields, html: "x" },
    { revision: 1, fields, preview: "forbidden" },
    { revision: 1, fields: [...fields, fields[0]] },
  ];
  for (const field of [
    { ...fields[0], choices: [{ id: "a", label: "A", upload_id: "forbidden" }] },
    { ...fields[0], max: 101 },
    { ...fields[0], min: 0.5 },
    {
      ...fields[0],
      choices: [
        { id: "a", label: "A" },
        { id: "a", label: "B" },
      ],
    },
    { ...fields[1], default: null },
    { ...fields[1], default: 1.25 },
    { ...fields[1], step: 0 },
    { ...fields[1], step: 1e-20 },
    { ...fields[1], max: Infinity },
    { ...fields[2], min: 0 },
    { ...fields[2], dynamic: false },
    { ...fields[3], choices: [] },
    { ...fields[3], default: "missing" },
    { ...fields[3], choices: [{ id: "a", label: "A", path: "/tmp/a" }] },
  ])
    bad.push({ revision: 1, fields: [field] } as any);
  for (const v of bad)
    assert.equal(validateGalleryActionSchema(v, allowed), undefined, JSON.stringify(v));
});
test("a failed open leaves a way to start a new panel", () => {
  const panel = readFileSync(
    new URL("../components/outputs/GalleryActionPanel.svelte", import.meta.url),
    "utf8",
  );
  assert.match(panel, /!submission && \(!!error \|\| state==="failed"\)/u);
});
test("dynamic image pages can grow beyond 100 while selection remains bounded", () => {
  const images = Array.from({ length: 229 }, (_, index) => ({
    id: `choice-${index}`,
    label: `Choice ${index}`,
  }));
  const schema = validateGalleryActionSchema({
    revision: 1,
    fields: [{ ...fields[0], choices: images }],
  });
  assert.equal(schema?.fields[0]?.choices.length, 229);
  assert.equal(
    validateGalleryActionSchema({
      revision: 1,
      fields: [{ ...fields[0], dynamic: false, choices: images }],
    }),
    undefined,
  );
});
test("effective values require exact keys, finite aligned numbers and unique scoped choices", () => {
  const schema = validateGalleryActionSchema({ revision: 1, fields }, allowed)!;
  const values = { images: ["a"], amount: 1.5, confirm: true, format: "png" };
  assert.equal(validateGalleryActionValues(schema, values), undefined);
  for (const patch of [
    { amount: 1.25 },
    { amount: NaN },
    { images: ["a", "a"] },
    { images: ["foreign"] },
    { extra: true },
    { format: "foreign" },
    { confirm: null },
  ])
    assert.ok(validateGalleryActionValues(schema, { ...values, ...patch } as any));
  assert.ok(validateGalleryActionValues(schema, null as any));
});
